ledger/cryptarchia-ledger/src/lib.rs

```diff
@@ -741,4 +742,82 @@ pub mod tests {
             Err(LedgerError::NullifierExists)
         ));
     }
+
+    #[test]
+    fn test_conversions_for_leader_proof() {
+        let commitment = Commitment::from([0u8; 32]);
+        let commitment_bytes: [u8; 32] = commitment.into();
+
+        let _zero_bytes = [0u8; 32];
+        assert!(matches!(commitment_bytes, _zero_bytes));
+
+        let commitment_ref = commitment.as_ref();
+        assert_eq!(commitment_ref, &_zero_bytes);
+
+        let nullifier = Nullifier::from([0u8; 32]);
+        let _nullifier_bytes: [u8; 32] = nullifier.into();
+        assert!(matches!(_nullifier_bytes, _zero_bytes));
+
+        let slot = Slot::genesis();
+        let leader_proof = LeaderProof::dummy(slot, commitment, Commitment::from([0; 32]));
+
+        assert_eq!(leader_proof.commitment(), &commitment);
+        assert_eq!(leader_proof.evolved_commitment(), &commitment);
+        assert_eq!(leader_proof.nullifier(), &nullifier);
+
+        // Test ser/de of compact representation for Nullifier
+        assert_tokens(&nullifier.compact(), &[Token::BorrowedBytes(&[0; 32])]);
+
+        // Test ser/de of compact representation for Commitment
+        assert_tokens(&commitment.compact(), &[Token::BorrowedBytes(&[0; 32])]);
+    }
+
+    #[test]
+    fn test_cryptarchia_ledger_error_cases() {
+        let coin = coin(0);
+        let commitment = coin.commitment();
+        let (ledger, genesis) = ledger(&[commitment]);
+
+        let ledger_state = ledger.state(&genesis).unwrap().clone();
+        let ledger_config = ledger.config();
+
+        let slot = Slot::genesis() + 10;
+        let ledger_state2 = ledger_state
+            .update_epoch_state::<HeaderId>(slot, ledger_config)
+            .expect("Ledger needs to move forward");
+
+        let slot2 = Slot::genesis() + 1;
+        let update_epoch_err = ledger_state2
+            .update_epoch_state::<HeaderId>(slot2, ledger_config)
+            .err();
+
+        // Time cannot flow backwards
+        match update_epoch_err {
+            Some(LedgerError::InvalidSlot { parent, block })
+                if parent == slot && block == slot2 => {}
+            _ => panic!("error does not match the LedgerError::InvalidSlot pattern"),
+        };
+
+        let ledger_state = ledger.state(&genesis).unwrap().clone();
+        let actual_slot = ledger_state.slot();
+        let proof = LeaderProof::dummy(actual_slot, commitment, commitment);
+        let epoch_state = ledger_state.epoch_state();
+
+        assert!(ledger_state.can_spend(&commitment));
+        assert_eq!(epoch_state.total_stake(), Value::from(1u32));
+        assert_eq!(coin.value(), Value::from(1u32));
+
+        assert!(ledger_state.can_lead(&commitment));
+        assert!(epoch_state.is_eligible_leader(&commitment));
+
+        let apply_proof_err = ledger_state
+            .try_apply_proof::<HeaderId>(&proof, ledger_config)
+            .err();
+
+        // Commitment cannot be spent twice
```

Comment:
nit: we're not trying to spend a coin twice here (that would result in a duplicate nullifier), but rather create a coin that has the same commitment as an existing one

Reply:
I see, perhaps I still don't understand what commitment is. I thought it is a pledge to spend a coin to be eligible to get voted as a leader? Some kind of a voting ticket ?

Reply:
The way you prove you are the leader is by proving ownership of a coin that won the lottery. The way you prove ownership of a coin is by spending it (prove you have the spending credentials). 
However, you still want the coin back after winning the election, so you send it back to yourself. The duplicate commitment error is because you're trying to send yourself the same coin that you spent while you're supposed to create a new one